vidarr-pluginapi/src/test/java/ca/on/oicr/gsi/vidarr/TaggedUnionInputTypeTest.java

```diff
@@ -0,0 +1,211 @@
+package ca.on.oicr.gsi.vidarr;
+
+import static ca.on.oicr.gsi.vidarr.VeryLongString.VERY_LONG;
+
+import ca.on.oicr.gsi.Pair;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.function.ThrowingRunnable;
+
+public class TaggedUnionInputTypeTest extends InputTypeTest {
+  String taggedUnionOneFieldStringFormat = "{\"is\":\"tagged-union\",\"options\":{\"%s\":\"%s\"}}";
+
+  @Override
+  public void testSerialize() {
+    for (Map.Entry<InputType, String> type : primitiveTypes.entrySet()) {
+      serializeTester(
+          String.format(taggedUnionOneFieldStringFormat, "field", type.getValue()),
+          InputType.taggedUnion(new Pair<>("field", type.getKey())));
+    }
+  }
+
+  @Test
+  public void testSerializeMany() {
+    List<Pair<String, InputType>> varargs = new LinkedList<>();
+    ObjectNode root = MAPPER.createObjectNode(), options = MAPPER.createObjectNode();
+    root.set("is", MAPPER.convertValue("tagged-union", JsonNode.class));
+    String json = "";
+
+    // FIXME the limit is set to 'b' here because at 'c' it stops serializing in alphabetical order.
```

Comment:
@apmasell do you know why this might be happening? or how it could be avoided in testing? JSON is not supposed to be order sensitive but idk how to compare other than String compares

Reply:
There's probably a hash map somewhere in there. Try `MAPPER.configure(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS, true);`

Maybe don't compare them as strings. You could take `JsonNode` for the first argument to `serializeTester` and take the `InputType` and do something like `json.equals(MAPPER.readValue(MAPPER.writeValueAsString(inputType), JsonNode.class))`. That will compare them as raw JSON objects without involving the special deserializer.

Reply:
I have that configuration set at InputTypeTest:28. 
If we don't involve the special deserializer, then we're not really testing deserialization, are we? I will think about how to rework this test

Reply:
I guess, there's two paths here that prove that we can round trip through the serializer:
- string of json -> InputType -> string of json
- InputType -> string of json -> InputType

We don't want to guarantee that the two strings in the first path are equal as strings. We care mostly that the second path is produces a functionally equivalent object after deserialisation.

Reply:
My concern about that is that if serialization produces erroneous json, but deserialization accepts erroneous json, then that round trip is going to work. If it's never compared to good json then it may cause issues if something else needs to handle the json output. But if this is the only way to test, then sure.

I'm confused why the equivalent test in ObjectInputTypeTest works all the way up to 'z' though.

Reply:
I looked at the Map types in this test for TaggedUnion and Object differ. They are both using TreeMaps, however TaggedUnion does put its TreeMap through `Collections.unmodifiableMap`, could this cause the difference in output?

Reply:
It shouldn't. The `unmodfiableMap` method doesn't change iteration order or anything: https://github.com/openjdk-mirror/jdk7u-jdk/blob/master/src/share/classes/java/util/Collections.java#L1324

I think your concern for erroneous JSON can be done with another test that checks that known JSON matches a constructed object. I don't know enough about what's going on inside Jackson.